src/celeritas/optical/PhysicsParams.cc

```diff
@@ -0,0 +1,156 @@
+ * physics data storage.
+ */
+PhysicsParams::PhysicsParams(Input input)
+{
+    CELER_EXPECT(input.action_registry);
+
+    // Create actions and register them
+    {
+        auto& action_reg = *input.action_registry;
+
+        // Discrete selection action
+        discrete_action_
+            = std::make_shared<DiscreteSelectAction>(action_reg.next_id());
+        action_reg.insert(discrete_action_);
+
+        // Create models
+        models_ = this->build_models(input.model_builders, action_reg);
+    }
+
+    // Construct data
+    HostValue host_data;
+    host_data.scalars.num_models = this->num_models();
+    this->build_options(input.options, host_data);
+    this->build_mfps(host_data);
+
+    // Copy data to device
+    data_ = CollectionMirror<PhysicsParamsData>{std::move(host_data)};
+
+    CELER_ENSURE(discrete_action_->action_id()
+                 == host_ref().scalars.discrete_action());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Builds optical models from list of model builders.
+ *
+ * Models are created and registered in the action registry.
+ */
+auto PhysicsParams::build_models(std::vector<SPConstModelBuilder> const& builders,
+                                 ActionRegistry& action_reg) const -> VecModels
+{
+    VecModels models;
+    models.reserve(builders.size());
+
+    for (auto const& build : builders)
+    {
+        auto action_id = action_reg.next_id();
+        auto model = (*build)(action_id);
+
+        CELER_ASSERT(model);
+        CELER_ASSERT(model->action_id() == action_id);
+
+        action_reg.insert(model);
+        models.push_back(std::move(model));
+    }
+
+    CELER_ENSURE(models.size() == builders.size());
+    return models;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Builds optical physics options.
+ */
+void PhysicsParams::build_options(Options const& /* options */,
+                                  HostValue& /* host_data */) const
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Builds MFP tables for each optical model.
+ *
+ * Iterates over the constructed optical models and has them build their MFP
+ * grids for each optical material. These grids are then mapped to per-model
+ * value tables, and the mappings are stored in the optical PhysicsData.
+ */
+void PhysicsParams::build_mfps(HostValue& data) const
+{
+    auto build_grid_id = make_builder(&data.grid_ids);
+    auto build_table = make_builder(&data.tables);
+    auto build_model_tables = make_builder(&data.model_tables);
+
+    for (auto model_id : range(ModelId{models_.size()}))
+    {
+        auto const& model = *models_[model_id.get()];
+
+        // Build the per material grids
+        GenericGridInserter inserter(&data.reals, &data.grids);
+        std::vector<ValueGridId> grid_ids;
+
+        model.build_mfps(detail::MfpBuilder{&inserter, &grid_ids});
```

Comment:
The grid inserter should be outside this loop so that it can deduplicate data across models. I think we can also move most of this function into the implementation of `MfpBuilder`, which could include all the builders: see `celeritas/optical/detail/MatScintSpecInserter.hh`.